File: helpers/r_helpers/make_stimtimes.R
make_stimtimes = function(stim_nodes, args_dict){
  
  # """
  #   Creates task timing and timeseries for all nodes in network
  #   Parameters specified in args_dict:
  #       Tmax = task length
  #       dt = sampling rate
  #       stim_nodes = nodes that are stimulated by the task
  #       tasktiming = block task array is created if not specified. 
  #       ncommunities = number of communities in network
  #       nodespercommunity = number of nodes per community in network
  #       sa = start point of stimulation
  #       ea = end point of stimulation
  #       iv = interstim interval
  #       W = adjacency matrix (prioritized over ncommunities and nodespercommunity)
  #   Returns: 
  #       tasktiming = 1D array of resampled tasktiming depending on dt 
  #       stimtimes = 2D array with nodes in rows and time points in columns
  #   """
  
  # Initialize parameters
  Tmax=args_dict$Tmax
  dt=args_dict$dt
  stim_mag=args_dict$stim_mag
  
  # Check if there is an adjacency matrix specified 
  if(is.null(args_dict$W)){
    ncommunities = args_dict$ncommunities
    nodespercommunity = args_dict$nodespercommunity
    totalnodes = nodespercommunity*ncommunities
  } else {
    totalnodes = dim(args_dict$W)[1]
  }
  
  # resampled task length depending on sampling rate dt
  # TT = seq(1,Tmax,dt)
  TT = seq(0,Tmax,dt)
  
  
  # Check if there is a taasktiming specified in the inputs 
  tasktiming=args_dict$tasktiming
  if (is.null(tasktiming)){    
    # If not task specified create task timing
    tasktiming = rep(0, length(TT))
    sa = args_dict$sa
    ea = args_dict$ea
    iv = args_dict$iv
    for (t in 1:(length(TT))){
      if (t%%iv>sa & t%%iv<ea){
        tasktiming[t] = 1.0
      }
    }
  }else if(length(tasktiming) != length(TT)){
    short_tasktiming = tasktiming
    tasktiming = rep(NA, length(TT))
    cur_t = 1
    
    # resample input tasktiming to Tmax/dt
    for(i in 1:length(short_tasktiming)){
      while(floor(cur_t)<length(short_tasktiming)){
        cur_stim = short_tasktiming[floor(cur_t)]
        first_na_index = which(is.na(tasktiming))[1]
        tasktiming[first_na_index] = cur_stim
        cur_t = cur_t + dt
      }
    }
    tasktiming[which(is.na(tasktiming))] = 0
  }
  
  stimtimes = matrix(0, totalnodes, length(TT))
  # When task is ON the activity for a stim_node at that time point changes the size of stim_mag
  for (t in 1:length(TT)){
    if (tasktiming[t] == 1){
      stimtimes[stim_nodes,t] = stim_mag
    }
  }
  
  return(list(tasktiming = tasktiming, stimtimes = stimtimes))
}

File: helpers/make_stimtimes.py
import numpy as np
import pandas as pd

from make_network import default_args
        
def make_stimtimes(stim_nodes, args_dict = default_args):
    
    """
    Creates task timing and timeseries for all nodes in network

    Parameters:
        Tmax = task length
        dt = sampling rate
        stim_nodes = nodes that are stimulated by the task
        tasktiming = block task array is created if not specified. If specified must be of length Tmax/dt
        ncommunities = number of communities in network
        nodespercommunity = number of nodes per community in network
        on_len = duration of stimulus
        off_len = duration of off period

    Returns: 
        2D array with nodes in rows and time points in columns
    """
    
    # Initialize parameters
    Tmax=args_dict['Tmax']
    dt=args_dict['dt']
    stim_mag=args_dict['stim_mag']
    tasktiming=args_dict['tasktiming']
    ncommunities = args_dict['ncommunities']
    nodespercommunity = args_dict['nodespercommunity']
    on_len = int(np.ceil(args_dict['on_len']/dt))
    off_len = int(np.ceil(args_dict['off_len']/dt))
    alternate_stim_nodes = args_dict['alternate_stim_nodes']
    
    totalnodes = nodespercommunity*ncommunities
    
    T = np.arange(0,Tmax,dt)
    num_blocks = int(np.ceil(len(T)/(on_len + 2*off_len)))
    
    # Construct timing 
    if tasktiming is None:
        tasktiming = np.tile(np.concatenate([np.zeros(off_len), np.ones(on_len), np.zeros(off_len)]), num_blocks)
    
    # If tasktiming ended up being longer than T
    if len(tasktiming) > len(T):
        T = np.arange(0,len(tasktiming)*dt,dt)
        args_dict.update({'Tmax': len(T)})
    
    stimtimes = np.zeros((totalnodes,len(T)))
    
    # When task is ON the activity for a stim_node at that time point changes the size of stim_mag
    for t in range(len(T)):
        if tasktiming[t] == 1:
            stimtimes[stim_nodes,t] = stim_mag
               
    if alternate_stim_nodes:
        # Calculate number of blocks for each node if you have alternating stim nodes
        num_node_blocks = np.floor(len(T)/(on_len + 2*off_len)/len(stim_nodes))
        node_stim_dur = int(num_node_blocks*(on_len + 2*off_len))

        # Make sure the tasktiming and stimtime length are the same
        new_task_len = int((node_stim_dur)*len(stim_nodes))
        tasktiming = tasktiming[:new_task_len]
        stimtimes = np.array([x[:new_task_len] for x in stimtimes])

        # Turn other nodes off when one node is on
        for i in stim_nodes:
            off_nodes = [x for x in stim_nodes if x != i]
            off_range_start = int(i*node_stim_dur)
            off_range_end = int(node_stim_dur*(i+1)-1)
            stimtimes[off_nodes, off_range_start:off_range_end] = 0 
            
    return(tasktiming, stimtimes, args_dict)